src/go/k8s/internal/controller/redpanda/redpanda_controller.go

```diff
@@ -531,390 +524,294 @@ func setHelmLabelsAndAnnotations(object client.Object, rp *v1alpha2.Redpanda) {
-		if apierrors.IsNotFound(err) {
-			repo = r.createHelmRepositoryFromTemplate(rp)
-			if errCreate := r.Client.Create(ctx, repo); errCreate != nil {
-				r.event(rp, rp.Status.LastAttemptedRevision, v1alpha2.EventSeverityError, fmt.Sprintf("error creating HelmRepository: %s", errCreate))
-				return rp, repo, fmt.Errorf("error creating HelmRepository: %w", errCreate)
+// based on https://github.com/fluxcd/pkg/blob/30c101fc7c9fac4d84937ff4890a3da46a9db2dd/ssa/manager_apply.go#L94-L140
+func (r *RedpandaReconciler) ServerSideApply(ctx context.Context, resource *unstructured.Unstructured) error {
+	log := ctrl.LoggerFrom(ctx).WithName("RedpandaReconciler.reconcile.ServerSideApply")
+	log.V(DebugLevel).Info("resource creation", "resource", resource)
+	existingObject := &unstructured.Unstructured{}
+	existingObject.SetGroupVersionKind(resource.GetObjectKind().GroupVersionKind())
+	getError := r.Get(ctx, client.ObjectKeyFromObject(resource), existingObject)
+
+	dryRunObject := resource.DeepCopy()
+	if err := r.dryRunApply(ctx, dryRunObject); err != nil {
+		if !k8serrors.IsNotFound(getError) && isImmutableError(err) {
+			log.V(DebugLevel).Info("immutable error", "resource", resource)
+			if err := r.Delete(ctx, existingObject, client.PropagationPolicy(metav1.DeletePropagationBackground)); err != nil && !k8serrors.IsNotFound(err) {
+				return err
 			}
-			r.event(rp, rp.Status.LastAttemptedRevision, v1alpha2.EventSeverityInfo, fmt.Sprintf("HelmRepository '%s/%s' created ", rp.Namespace, rp.GetHelmRepositoryName()))
-		} else {
-			r.event(rp, rp.Status.LastAttemptedRevision, v1alpha2.EventSeverityError, fmt.Sprintf("error getting HelmRepository: %s", err))
-			return rp, repo, fmt.Errorf("error getting HelmRepository: %w", err)
+			return r.ServerSideApply(ctx, resource)
 		}
-	}
-	rp.Status.HelmRepository = rp.GetHelmRepositoryName()
 
-	return rp, repo, nil
-}
-
-func (r *RedpandaReconciler) reconcileDelete(ctx context.Context, rp *v1alpha2.Redpanda) (ctrl.Result, error) {
-	if err := r.deleteHelmRelease(ctx, rp); err != nil {
-		return ctrl.Result{}, err
-	}
-	if controllerutil.ContainsFinalizer(rp, FinalizerKey) {
-		controllerutil.RemoveFinalizer(rp, FinalizerKey)
-		if err := r.Client.Update(ctx, rp); err != nil {
-			return ctrl.Result{}, err
-		}
+		return err
 	}
-	return ctrl.Result{}, nil
-}
 
-func (r *RedpandaReconciler) createHelmRelease(ctx context.Context, rp *v1alpha2.Redpanda) (*helmv2beta2.HelmRelease, error) {
-	// create helmRelease resource from template
-	hRelease, err := r.createHelmReleaseFromTemplate(ctx, rp)
-	if err != nil {
-		r.event(rp, rp.Status.LastAttemptedRevision, v1alpha2.EventSeverityError, fmt.Sprintf("could not create helm release template: %s", err))
-		return hRelease, fmt.Errorf("could not create HelmRelease template: %w", err)
+	// do not apply objects that have not drifted to avoid bumping the resource version
```

Comment:
Does server side apply not handle this for us? ☹️

Reply:
I could check that again. Let me modify the code a bit and verify with my test.